### PR TITLE
[opengl] Serialize ndarrays and ndarray-based kernels in AOT.

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #3556
* #3555
* #3554

Milestone: serialize a ndarray version of mpm88 with AOT module.

Serialized data structure is subject to change, will need to carefully
rethink about what's the minimal information should be included with an
ndarray kernel.

### CODE REVIEW

File: taichi/backends/opengl/aot_data.h [hunk 1]
```diff
@@ -43,14 +43,35 @@ struct CompiledFieldData {
             column_num);
 };
 
+struct CompiledNdarrayData {
+  std::string field_name;
+  uint32_t dtype;
+  std::string dtype_name;
+  size_t dim{0};
+  std::vector<int> shape;
```
Comment (k-ye) [inline taichi/backends/opengl/aot_data.h:51]: Actually we don't want to restrict the shape here?

Comment (ailzhang) [reply taichi/backends/opengl/aot_data.h:51]: Yup it's more like a FYI to show an example shape, `dim` info above is more important.

Comment (k-ye) [reply taichi/backends/opengl/aot_data.h:51]: Ack, let's just remove it :-)

**Review by qiao-bo — APPROVED** [review]
LGTM!

File: taichi/backends/opengl/aot_data.h (hunk 1 above)
Comment (k-ye) [reply taichi/backends/opengl/aot_data.h:51]: Can we remove `shape`?

**Review by k-ye — APPROVED** [review]
LGTM thanks!

**Pull request merged**